### DISCUSSION TITLE
[3 Novembre 2022] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#56] [Cache (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/01990f53-2d93-4287-9f1c-a4f994ab856f)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665956/AE.2021-22.-.Appello.S1.-.2022-11-03.-.Esercizio.cache.-.Scheda.di.calcolo.pdf)

**Comment by Frisbon** [comment]
### [Sto esercizio vale 11 punti ma si fa in meno di 20 minuti](https://github.com/sapienzastudentsnetwork/archelab2223/files/11891246/New.Section.2.3.pdf)

EDIT:
(Non ho considerato, m'é sfuggita l'ultima hit in basso a destra)

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/570aa766-fdbb-4ed7-83d2-d136d5dab5e7)

si riempie la tabella ottenendo

![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/a8f95ce8-86b0-4bde-bfaf-60ce6fab8d22)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1  + 1 + 27 + 4 * 4 * 8) * (3 * 2) \implies (29 + 128) * 6 \implies  157 * 6 \implies 942\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 22 + 64 * 4 * 8) * 8 \implies (23 + 2048) * 4 \implies  2071 * 4 \implies 8284\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 6 accessi in memoria, con tempo $\Large 300 * 6 = 1800\ \text{ns}$

- 3 accessi in L1, con tempo  $\Large 3 * 3 = 9\ \text{ns}$

- 3 accessi in L2, con tempo  $\Large 30 * 3 = 90\ \text{ns}$

$$
    \Large T_{tot} = 1800 + 9 + 90 = 1899\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{1899}{12} = 158.25\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{158.25 * 16}{4} = 633.00
$$

## 4) Word offset indirizzo 264 Cache L2

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (264 \gg 2)\ \\%\ 2^{6} \implies 66\ \\%\ 64 \implies 2
$$

## 5) Pagine Virtuali

ricavandosi la formula usando un approccio grafico

![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/89091a15-923c-4044-a0f3-9c20bbfe49c3)

si riempie la tabella ottenendo

![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/dca5b3b6-1234-4820-81b5-455c1fe5127d)

**Reply by matypist** [reply]
Confermo la proposta di soluzione di @FeddyLix17 , tutto chiaro e corretto 👍

**Comment by M1SC0** [comment]
![image](https://github.com/user-attachments/assets/b1da1114-9c3d-4521-af8a-efa3d020e3f0)
